gridded_etl_tools/utils/extractor.py

```diff
@@ -10,75 +10,64 @@
 if TYPE_CHECKING:  # pragma NO COVER
     from .. import dataset_manager
 
+from abc import ABC, abstractmethod
+import asyncio
+import nest_asyncio
 import pathlib
 import typing
 import ftplib
 import re
 import time
-import multiprocessing
 import logging
 
 
 log = logging.getLogger("extraction_logs")
 
 
-def pool(batch_processor: typing.Callable[..., bool], batch: typing.Sequence[typing.Sequence]) -> bool:
-    """
-    Launch a batch of requests simultaneously, wait for them all to complete, then return a boolean indicating
-    whether any of the requests were successful.
-
-    The batch processor is any generic callable that accepts the arguments provided in each entry of the batch.
-    Each entry in the batch is a list of arguments to be forwarded to the batch processor. The batch is submitted
-    to a thread pool which will call the batch processor on every entry of the batch, launching all the calls in
-    parallel and waiting for them to complete.
-
-    If any of the calls returns `True`, this will return `True`, otherwise it will return `False`.
-
-    Parameters
-    ----------
-    batch_processor
-        Function which will be called repeatedly, being passed a new entry in the batch each time it's called
-    batch
-        A 2D list of lists (or any sequential type) containing arguments to be forwarded to the given batch
-        processor
-
-    Returns
-    -------
-    bool
-        `True` if any of the batch calls returns `True`, `False` otherwise
-    """
-    # Success remains Trueuntil the first unsuccessful request is completed
-    success = True
+class Extractor(ABC):
+
+    def __init__(self, dm: dataset_manager.DatasetManager, concurrency_limit: int = 8):
```

Comment:
no default to number of cores - 1?

Reply:
Nope, the number of cores is completely irrelevant for this implementation. The main limitation is whether your target system gets angry at you. In the case of FTPs, that seems to happen for anything above 2. For other extractors, much higher might be possible

Reply:
For initial parses on Copernicus API, we want to immediately launch all thousands of requests so that they get registered in the queue. In that case, I'm guessing we'll need to override the default of 8 in ERA5's extract code.